tests/firedrake_adjoint/test_optimisation.py

```diff
@@ -157,13 +164,177 @@ def test_tao_simple_inversion(minimize, riesz_representation):
+        lam, = adj_inputs
+        return transform(lam, *self._args, **self._kwargs)
+
+    def recompute_component(self, inputs, block_variable, idx, prepared):
+        v, = inputs
+        return transform(v, *self._args, **self._kwargs)
+
+
+@pytest.mark.skipif(SLEPc is None, reason="SLEPc not available")
+@pytest.mark.parametrize("tao_type", ["lmvm",
+                                      "blmvm"])
+def test_simple_inversion_riesz_representation(tao_type):
+    """Test use of a Riesz map in inversion of source term in helmholze eqn
+    using TAO."""
+
+    riesz_representation = "L2"
+    mfn_parameters = {"mfn_type": "krylov",
+                      "mfn_tol": 1.0e-12}
+    tao_parameters = {"tao_type": tao_type,
+                      "tao_gatol": 1.0e-5,
+                      "tao_grtol": 0.0,
+                      "tao_gttol": 0.0,
+                      "tao_monitor": None}
+
+    pause_annotation()
+
+    mesh = UnitIntervalMesh(10)
+    V = FunctionSpace(mesh, "CG", 1)
+    source_ref = Function(V)
+    x = SpatialCoordinate(mesh)
+    source_ref.interpolate(cos(pi*x**2))
+    u_ref = _simple_helmholz_model(V, source_ref)
+
+    def forward(source):
+        c = Control(source)
+        u = _simple_helmholz_model(V, source)
+
+        J = assemble(1e6 * (u - u_ref)**2*dx)
+        rf = ReducedFunctional(J, c)
+        return rf
+
+    get_working_tape().clear_tape()
+    source = Function(V)
+    continue_annotation()
+    rf = forward(source)
+    pause_annotation()
+
+    solver = TAOSolver(
+        MinimizationProblem(rf), tao_parameters,
+        convert_options={"riesz_representation": riesz_representation})
+    x = solver.solve()
+    assert_allclose(x.dat.data, source_ref.dat.data, rtol=1e-2)
+
+    def forward_transform(source):
+        c = Control(source)
+        source = transform(source, TransformType.PRIMAL,
+                           riesz_representation,
+                           mfn_parameters=mfn_parameters)
+        u = _simple_helmholz_model(V, source)
+
+        J = assemble(1e6 * (u - u_ref)**2*dx)
+        rf = ReducedFunctional(J, c)
+        return rf
+
+    get_working_tape().clear_tape()
+    source_transform = transform(Function(V), TransformType.DUAL,
+                                 riesz_representation,
+                                 mfn_parameters=mfn_parameters)
+    continue_annotation()
+    rf_transform = forward_transform(source_transform)
+    pause_annotation()
+
+    solver_transform = TAOSolver(
+        MinimizationProblem(rf_transform), tao_parameters,
+        convert_options={"riesz_representation": "l2"})
+    x_transform = transform(solver_transform.solve(), TransformType.PRIMAL,
+                            riesz_representation,
+                            mfn_parameters=mfn_parameters)
+    assert_allclose(x_transform.dat.data, source_ref.dat.data, rtol=1e-2)
+
+    assert solver_transform.tao.getIterationNumber() <= solver.tao.getIterationNumber()
```

Comment:
Let me see if I understood right. This test considers the case where the optimisation solver uses the derivatives instead of the gradients. Then, `transform` (using mass Matrix) is applied to convert the `TaoSolver` results to the primal space. Is that right? If yes, That is extendable to other `tao_type`s, right?

Reply:
Short answer: `transform` applies a mass matrix inverse square root, so that the control is expanded in an $L^2$ orthonormal basis, and so that the $l^2$ inner product is the same as the $L^2$ inner product. This means the derivative is the same as the gradient, and we get mesh-independent optimizers using $l^2$ inner products. There's a minor hack where the dofs in the new basis are stored in `Function`s/`Cofunction`s.

Long answer: Assuming $L^2$ is the relevant inner product, and considering real spaces.

FEM encounters issues with Riesz maps because we choose a basis with local support, but sacrifice $L^2$ orthogonality. There's nothing to stop us choosing a basis which is $L^2$ orthogonal, it's just that the basis no longer has local support (in general).

However we can still transform from an expansion in the usual FEM basis to an expansion in an $L^2$ orthonormal basis,

$$\hat{\phi} = J \tilde{\phi},$$

where overhat means dofs in some $L^2$ orthonormal basis and tilde means dofs in the original FEM basis. We want

$$\hat{\phi}_2^T \hat{\phi}_1 = \tilde{\phi}_2^T J^T J \tilde{\phi}_1 = \tilde{\phi}_2^T M \tilde{\phi}_1,$$

for any two $\tilde{\phi}_1$, $\tilde{\phi}_2$, where $M$ is the mass matrix. This means $J$ needs to satisfy

$$J^T J = M.$$

So we can e.g. choose $J$ to be the principal square root of the mass matrix, or the upper Cholesky factor, or make other choices -- each corresponding to a possibly different choice of $L^2$ orthonormal basis.

Fortunately, as usual, we don't need $J$, only it's action, which we can compute matrix free. The test computes the matrix-free square root action using  SLEPc.